src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -991,6 +991,21 @@ public async Task<ImmutableArray<SourceGeneratedDocumentState>> GetSourceGenerat
                 return compilationInfo.GeneratedDocuments;
             }
 
+            public SourceGeneratedDocumentState? TryGetSourceGeneratedDocumentForAlreadyGeneratedId(DocumentId documentId)
+            {
+                var state = ReadState();
+
+                // If we are in FinalState, then we have correctly ran generators and then know the final contents of the
+                // Compilation. The GeneratedDocuments can be filled for intermediate states, but those aren't guaranteed to be
+                // correct and can be re-ran later.
+                if (state is FinalState finalState)
```

Comment:
so, if the state is FinalState are we guaranteed to not let go of that?  just trying to keep lifetimes around in my head.

Reply:
I believe that's the case; I'll add an assert, which is also generally good for understanding this whole thing.